c2rust-analyze/src/panic_detail.rs

```diff
@@ -0,0 +1,205 @@
+use backtrace::Backtrace;
+use log::warn;
+use rustc_span::{Span, DUMMY_SP};
+use std::any::Any;
+use std::cell::Cell;
+use std::fmt::Write as _;
+use std::panic::{self, PanicInfo, UnwindSafe};
+
+/// Detailed information about a panic.
+#[derive(Clone, Debug)]
+pub struct PanicDetail {
+    msg: String,
+    loc: Option<String>,
+    relevant_loc: Option<String>,
+    backtrace: Option<Backtrace>,
+    span: Span,
+}
+
+impl PanicDetail {
+    /// Create a new `PanicDetail` containing only a message, with no location or backtrace
+    /// information.
+    pub fn new(msg: String) -> PanicDetail {
+        PanicDetail {
+            msg,
+            loc: None,
+            relevant_loc: None,
+            backtrace: None,
+            span: DUMMY_SP,
+        }
+    }
+
+    /// Returns `true` if this `PanicDetail` contains a backtrace.
+    pub fn has_backtrace(&self) -> bool {
+        self.backtrace.is_some()
+    }
+
+    /// Return a short (usually one-line) description of this panic.
+    pub fn to_string_short(&self) -> String {
+        let loc_str = self
+            .relevant_loc
+            .as_ref()
+            .or(self.loc.as_ref())
+            .map_or("[unknown]", |s| &*s);
+        format!("{}: {}", loc_str, self.msg.trim())
+    }
+
+    /// Return a full description of this panic, including a complete backtrace if available.
+    pub fn to_string_full(&self) -> String {
+        let mut s = String::new();
+        let loc_str = self.loc.as_ref().map_or("[unknown]", |s| &*s);
```

Comment:
```suggestion
            .map_or("[unknown]", |s| s);
        format!("{}: {}", loc_str, self.msg.trim())
    }

    /// Return a full description of this panic, including a complete backtrace if available.
    pub fn to_string_full(&self) -> String {
        let mut s = String::new();
        let loc_str = self.loc.as_ref().map_or("[unknown]", |s| s);
```

A `clippy::borrow_deref_ref` warning.  The `&*` isn't necessary for a reborrow.